File: backend/app/schemas/analytics.py
from datetime import date

from pydantic import BaseModel

from app.models.enums import GameType, PeriodType


class PlayerStatsQuery(BaseModel):
    game: GameType
    period_type: PeriodType = PeriodType.ALL_TIME
    tournament_id: str | None = None


class PlayerStatsResponse(BaseModel):
    player_id: str
    in_game_name: str
    game: GameType
    period_type: PeriodType
    period_date: date
    matches_played: int
    matches_won: int
    games_played: int
    games_won: int
    total_kills: int
    total_deaths: int
    total_assists: int
    avg_kda: float
    win_rate: float
    most_played_agent: str | None
    agent_breakdown: dict | None

    model_config = {"from_attributes": True}


class TeamStatsResponse(BaseModel):
    team_id: str
    team_name: str
    game: GameType
    matches_played: int
    wins: int
    losses: int
    win_rate: float
    current_streak: int
    best_win_streak: int
    map_win_rates: dict | None

    model_config = {"from_attributes": True}


class MapStatsResponse(BaseModel):
    map_id: str
    map_name: str
    game: GameType
    total_games: int
    attack_side_wins: int
    defense_side_wins: int
    attack_win_rate: float
    avg_duration_seconds: float | None
    round_distribution: dict | None

    model_config = {"from_attributes": True}


class CompositionStatsResponse(BaseModel):
    composition: list[str]
    games_played: int
    wins: int
    win_rate: float
    avg_kills: float | None
    avg_deaths: float | None

    model_config = {"from_attributes": True}


class TournamentSummaryResponse(BaseModel):
    tournament_id: str
    tournament_name: str
    game: GameType
    total_matches: int
    completed_matches: int
    total_teams: int
    top_teams: list[dict]
    top_players_kda: list[dict]
    most_played_map: str | None
    avg_match_duration_seconds: float | None


class RankingEntry(BaseModel):
    rank_position: int
    team_id: str
    team_name: str
    team_tag: str
    team_logo_url: str | None
    points: int
    wins: int
    losses: int
    game_wins: int
    game_losses: int
    win_rate: float

    model_config = {"from_attributes": True}

File: frontend/src/types/analytics.ts
import type { GameType } from "./tournament";

export interface PlayerStats {
  player_id: string;
  in_game_name: string;
  game: GameType;
  period_type: string;
  period_date: string;
  matches_played: number;
  matches_won: number;
  games_played: number;
  games_won: number;
  total_kills: number;
  total_deaths: number;
  total_assists: number;
  avg_kda: number;
  win_rate: number;
  most_played_agent: string | null;
  agent_breakdown: Record<string, { games: number; wins: number; kda: number }> | null;
}

export interface MapStats {
  map_id: string;
  map_name: string;
  game: GameType;
  total_games: number;
  /** 平均ラウンド差（小さいほど競った試合が多いMAP） */
  avg_round_margin: number | null;
  /** 2ラウンド差以内で決着した割合 */
  close_game_rate: number;
  avg_duration_seconds: number | null;
  round_distribution: Record<string, number> | null;
}

export interface CompositionStats {
  composition: string[];
  games_played: number;
  wins: number;
  win_rate: number;
  avg_kills: number | null;
  avg_deaths: number | null;
}

export interface RankingEntry {
  rank_position: number;
  team_id: string;
  team_name: string;
  team_tag: string;
  team_logo_url: string | null;
  points: number;
  wins: number;
  losses: number;
  game_wins: number;
  game_losses: number;
  win_rate: number;
}

export interface TournamentSummary {
  tournament_id: string;
  tournament_name: string;
  game: GameType;
  total_matches: number;
  completed_matches: number;
  total_teams: number;
  top_teams: Array<{ team_id: string; team_name: string; wins: number }>;
  top_players_kda: Array<{ player_id: string; avg_kda: number }>;
  most_played_map: string | null;
  avg_match_duration_seconds: number | null;
}

// ── 新規追加型 ─────────────────────────────────────────────────────────────────

export interface OverviewKpi {
  total_matches: number;
  total_games: number;
  total_tournaments: number;
  /** チームを指定したときのみ算出される。全体表示では null */
  overall_win_rate: number | null;
  /** 平均ラウンド差。閲覧者の視点に依存しない競り合いの指標 */
  avg_round_margin: number | null;
  avg_match_duration_seconds: number | null;
  most_played_map: string | null;
  most_played_agent: string | null;
  active_teams: number;
  active_players: number;
}

export interface TrendPoint {
  date: string;
  matches: number;
  /** 平均ラウンド差。勝率は視点依存で意味を持たないため置き換えた */
  avg_round_margin: number | null;
  avg_kda: number;
  avg_duration_seconds: number | null;
}

export interface AgentStat {
  agent: string;
  games: number;
  wins: number;
  losses: number;
  win_rate: number;
  pick_rate: number;
  ban_rate: number;
  avg_kda: number;
  avg_kills: number;
  avg_deaths: number;
  avg_assists: number;
}

export interface PlayerRankingEntry {
  rank: number;
  player_id: string;
  player_name: string;
  in_game_name: string | null;
  team_name: string | null;
  team_tag: string | null;
  avatar_url: string | null;
  games: number;
  win_rate: number;
  avg_kda: number;
  avg_kills: number;
  avg_deaths: number;
  avg_assists: number;
  most_played_agent: string | null;
}

export interface HeatMapCell {
  map_name: string;
  agent: string;
  win_rate: number;
  games: number;
}

export interface AnalyticsWinRate {
  overview: OverviewKpi;
  by_map: MapStats[];
  by_agent: AgentStat[];
}

export type AnalyticsPeriod = "7d" | "30d" | "90d" | "all";
